cleanrl/sac_atari.py

```diff
@@ -0,0 +1,339 @@
+    parser.add_argument("--total-timesteps", type=int, default=1000000,
+        help="total timesteps of the experiments")
+    parser.add_argument("--buffer-size", type=int, default=int(1e6),
+        help="the replay memory buffer size") # smaller than in original paper but evaluation is done only for 100k steps anyway
+    parser.add_argument("--gamma", type=float, default=0.99,
+        help="the discount factor gamma")
+    parser.add_argument("--tau", type=float, default=1.0,
+        help="target smoothing coefficient (default: 1)") # Default is 1 to perform replacement update
+    parser.add_argument("--batch-size", type=int, default=64,
+        help="the batch size of sample from the reply memory")
+    parser.add_argument("--learning-starts", type=int, default=2e4,
+        help="timestep to start learning")
+    parser.add_argument("--policy-lr", type=float, default=3e-4,
+        help="the learning rate of the policy network optimizer")
+    parser.add_argument("--q-lr", type=float, default=3e-4,
+        help="the learning rate of the Q network network optimizer")
+    parser.add_argument("--update-frequency", type=int, default=4,
+        help="the frequency of training updates")
+    parser.add_argument("--target-network-frequency", type=int, default=8000,
+        help="the frequency of updates for the target networks")
+    parser.add_argument("--alpha", type=float, default=0.2,
+        help="Entropy regularization coefficient.")
+    parser.add_argument("--autotune", type=lambda x:bool(strtobool(x)), default=True, nargs="?", const=True,
+        help="automatic tuning of the entropy coefficient")
+    parser.add_argument("--target-entropy-scale", type=float, default=0.88,
+        help="coefficient for scaling the autotune entropy target")
+    args = parser.parse_args()
+    # fmt: on
+    return args
+
+
+def make_env(env_id, seed, idx, capture_video, run_name):
+    def thunk():
+        env = gym.make(env_id)
+        env = gym.wrappers.RecordEpisodeStatistics(env)
+        if capture_video:
+            if idx == 0:
+                env = gym.wrappers.RecordVideo(env, f"videos/{run_name}")
+        env = NoopResetEnv(env, noop_max=30)
+        env = MaxAndSkipEnv(env, skip=4)
+        env = EpisodicLifeEnv(env)
+        if "FIRE" in env.unwrapped.get_action_meanings():
+            env = FireResetEnv(env)
+        env = ClipRewardEnv(env)
+        env = gym.wrappers.ResizeObservation(env, (84, 84))
+        env = gym.wrappers.GrayScaleObservation(env)
+        env = gym.wrappers.FrameStack(env, 4)
+        env.seed(seed)
+        env.action_space.seed(seed)
+        env.observation_space.seed(seed)
+        return env
+
+    return thunk
+
+
+def layer_init(layer, bias_const=0.0):
+    nn.init.kaiming_normal_(layer.weight)
+    torch.nn.init.constant_(layer.bias, bias_const)
+    return layer
+
+
+# ALGO LOGIC: initialize agent here:
+# NOTE: Sharing a CNN encoder between Actor and Critics is not recommended for SAC
```

Comment:
Don't they actually share the CNN encoder between critics and actor, but they stop the gradient going into the encoder from the actor's `fc1` and `fc_logits` though ?

Here is where they actually share the CNN between the Actor and Critics, if I am not mistaken.
https://github.com/denisyarats/pytorch_sac_ae/blob/7fa560e21c026c04bb8dcd72959ecf4e3424476c/sac_ae.py#L247

And here is where they detach the gradient of the actor layers so that it does not go into the shared CNN encoder:
https://github.com/denisyarats/pytorch_sac_ae/blob/7fa560e21c026c04bb8dcd72959ecf4e3424476c/sac_ae.py#L162

So I would say that the problem isn't really sharing the CNN encoder, but letting actor gradient backpropagate into the latter, as mentioned in the NOTE 2 lines below.

Since it seems to work in this implementation anyway, it should be fine to move ahead and merge.
It could be interesting to see how it fares with shared CNN encoder but stop_gradient in the actor level though.

Reply:
From the paper, Appendix B.2 for reference:
> The actor and critic networks both have separate encoders, although we share the weights of the conv layers between them. Furthermore, only the critic optimizer is allowed to update these weights (e.g. we truncate the gradients from the actor before they propagate to the shared conv layers).